atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/async/AsyncSessionManager.java

```diff
@@ -0,0 +1,236 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.keyvalue.cassandra.async;
+
+import java.net.InetSocketAddress;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicReference;
+
+import javax.net.ssl.SSLContext;
+
+import org.immutables.value.Value;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.datastax.driver.core.Cluster;
+import com.datastax.driver.core.HostDistance;
+import com.datastax.driver.core.PoolingOptions;
+import com.datastax.driver.core.ProtocolOptions;
+import com.datastax.driver.core.QueryOptions;
+import com.datastax.driver.core.RemoteEndpointAwareJdkSSLOptions;
+import com.datastax.driver.core.ThreadingOptions;
+import com.datastax.driver.core.exceptions.NoHostAvailableException;
+import com.datastax.driver.core.policies.AddressTranslator;
+import com.datastax.driver.core.policies.DefaultRetryPolicy;
+import com.datastax.driver.core.policies.LatencyAwarePolicy;
+import com.datastax.driver.core.policies.LoadBalancingPolicy;
+import com.datastax.driver.core.policies.RoundRobinPolicy;
+import com.datastax.driver.core.policies.TokenAwarePolicy;
+import com.datastax.driver.core.policies.WhiteListPolicy;
+import com.github.benmanes.caffeine.cache.Cache;
+import com.github.benmanes.caffeine.cache.Caffeine;
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import com.palantir.atlasdb.cassandra.CassandraKeyValueServiceConfig;
+import com.palantir.common.base.Throwables;
+import com.palantir.conjure.java.config.ssl.SslSocketFactories;
+import com.palantir.logsafe.Preconditions;
+import com.palantir.logsafe.SafeArg;
+import com.palantir.tritium.metrics.registry.TaggedMetricRegistry;
+
+public final class AsyncSessionManager {
+
+    @Value.Immutable
+    interface UniqueCassandraCluster {
+        @Value.Parameter
+        Set<InetSocketAddress> servers();
+
+    }
+
+    static class SimpleAddressTranslator implements AddressTranslator {
+
+        private final Map<String, InetSocketAddress> mapper;
+
+        SimpleAddressTranslator(CassandraKeyValueServiceConfig config) {
+            this.mapper = config.addressTranslation();
+        }
+
+        @Override
+        public void init(Cluster cluster) {
+        }
+
+        @Override
+        public InetSocketAddress translate(InetSocketAddress address) {
+            InetSocketAddress temp = mapper.getOrDefault(address.getHostString(), address);
+            return new InetSocketAddress(temp.getAddress(), temp.getPort());
+        }
+
+        @Override
+        public void close() {
+        }
+    }
+
+    private static final Logger log = LoggerFactory.getLogger(AsyncSessionManager.class);
+    private static final AtomicReference<AsyncSessionManager> FACTORY = new AtomicReference<>(null);
```

Comment:
2 things:

Did we check sessions can switch keyspaces dynamically (i.e. on every query)? If yes that's great.

Regardless if that is the case, we should have the default implementation still create a a full blown CQL client per TransactionManager, but provide a way for internal products to just provide the KVS that should be used. Similar to how the ability to override the TimestampCache was recently added: https://github.com/palantir/atlasdb/pull/4199/files#diff-b89000610cc473386808b5a80a1be1e9R270

Products that start multiple TransactionManagers will have appropriate ways to correctly manage the lifetime, without using statics and just provide an implementation of this.

Reply:
Per docs for the driver currently used, having sessions which are not keyspace bound is preferable as there are [known limitations](https://docs.datastax.com/en/developer/java-driver/3.7/manual/#session) to having multiple sessions.